---
title: "Introduction to single-cell RNA-seq analysis"
subtitle: 'Normalisation - GSM3872434 sample'

author: "Stephane Ballereau/ Kamal Kishore"
output:
  html_document:
    df_print: paged
    toc: yes
    number_sections: true
    code_folding: show
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: true
  html_book:
    code_folding: show
---

# Normalisation - Exercises

Exercise: apply the deconvolution normalisation method on a
single sample: ETV6-RUNX1_1 (aka GSM3872434).

```{r norm_Caron.knitr_options, echo=FALSE, results="hide", message=FALSE}
require(knitr)
# set working directory# set working directory# set working directory
knitr::opts_knit$set(root.dir = '/Users/kishor01/Desktop/SingleCellCourseCRUK/Bitesize_SingleCell/CourseMaterials/Exercises')
# set other chunk variables
opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, eval=FALSE)
opts_chunk$set(fig.width=7, fig.height=7)
set.seed(123) # for reproducibility
```

```{r libraries_norm_GSM3872434_allCells, results='hide', message=FALSE, warning=FALSE}
library(scater)
library(scran)
library(tidyverse)
library(BiocSingular)
library(BiocParallel)

bpp <- MulticoreParam(7)
```

## Load object

We will load the R object created after QC.

```{r GSM3872434_readIn}
# Read object in:
# remember getwd() and dir()
sce <- readRDS("../Robjects/Caron_filtered_genes.rds")
colData(sce)$SampleName <- colData(sce)$Sample
```

Select cells for ETV6-RUNX1_1 sample:

```{r GSM3872434_downsample}
# have new list of cell barcodes for each sample
sce.master <- sce
vec.bc <- colData(sce.master) %>%
	data.frame() %>%
	filter(SampleName == "ETV6-RUNX1_1") %>%
	group_by(SampleName) %>%
	pull(Barcode)
```

Check the Number of cells in the sample:

```{r GSM3872434_barcodes}
table(colData(sce.master)$Barcode %in% vec.bc)
```

Subset the cells from the SCE object:

```{r GSM3872434_subset_sce}
tmpInd <- which(colData(sce.master)$Barcode %in% vec.bc)
sce <- sce.master[,tmpInd]
sce
```

Check columns data:

```{r GSM3872434_SampleName_table}
head(colData(sce))
table(colData(sce)$SampleName)
```

<!--
# mind that genes were filtered using all cells, not just those sampled here.
-->

## Exercise 1 : Deconvolution

Clusters of cells first identified to help form sensible pools of cells. 

### Cluster cells

```{r comp_quickClus_norm_GSM3872434_allCells}
set.seed(100) # clusters with PCA from irlba with approximation
clust <- quickCluster(sce) # slow with all cells.
table(clust)
```

### Compute size factors

Scaling factors are then computed from the identified cluster.


```{r calculateSumFactors_norm_GSM3872434_allCells}
# deconvolve


# set size factors


# size factors distribution summary


```

Check the relation between the deconvolution size factors against library size factors. To do this compute library size factors:

```{r scatter_deconvSf_libSf_prep_norm_GSM3872434_allCells}
# compute library size factors


# make data frame keeping library and deconvolution size factors for plotting


```


Generate a scatter plot of library size against deconvolution size factors:
 
```{r scatter_deconvSf_libSf_colBy_more_norm_GSM3872434_allCells}
# plot deconv.sf against lib.sf

# colour by library size


```

### Apply size factors

Apply the deconvolution size factors on the dataset:
<!--
For each cell, raw counts for genes are divided by the size factor for that cell
and log-transformed so downstream analyses focus on genes with strong relative
differences. We use `scater::logNormCounts()`.
-->

```{r logNormCounts_norm_GSM3872434_allCells}


```
