---
title: "CRUK CI Summer School 2022 - introduction to single-cell RNA-seq analysis"
subtitle: 'Dataset Integration'
author: "Abbi Edwards"
output:
  html_document:
    df_print: paged
    toc: yes
    number_sections: true
    code_folding: show
    css: ../css/boxes.css
---

# Data integration - PBMMC and ETV6-RUNX samples


## Learning Objectives

* Understand why/when batch correction is required
* Understand where data set integration fits in the workflow
* Understand one main method for batch correction
* Understand how to check your batch correction has worked

```{r library_PBMMC_ETV6-RUNX1, warning=FALSE, message=FALSE}
library(scater)
library(scran)
library(batchelor)
library(bluster)
library(tidyverse)
library(pheatmap)
library(clustree)
library(Cairo)
library(BiocSingular)
library(cowplot)
```

Source: ['Integrating Datasets'](https://osca.bioconductor.org/integrating-datasets.html) chapter in the OSCA book.

## Motivation

Large single-cell RNA sequencing (scRNA-seq) projects usually need to generate data across multiple batches due to logistical constraints. However, the processing of different batches is often subject to uncontrollable differences, e.g., changes in operator, differences in reagent quality. This results in systematic differences in the observed expression in cells from different batches, which we refer to as “batch effects”. Batch effects are problematic as they can be major drivers of heterogeneity in the data, masking the relevant biological differences and complicating interpretation of the results.

Computational removal of batch-to-batch variation allows us to combine data across multiple batches for a consolidated downstream analysis. However, existing methods based on linear models (Ritchie et al. 2015; Leek et al. 2012) assume that the composition of cell populations are either known or the same across batches. To overcome these limitations, bespoke methods have been developed for batch correction of single-cell data (Haghverdi et al. 2018; Butler et al. 2018; Lin et al. 2019) that do not require a priori knowledge about the composition of the population. This allows them to be used in workflows for exploratory analyses of scRNA-seq data where such knowledge is usually unavailable.

## Load the data

To demonstrate the code we will load the data we have for two samples which are technical replicates of one another from two different 10X runs and so if there were no batch effect they should be the same. These samples have been processed as discussed up until this point in the course.

```{r, loadSimpleSamples}

ori.sce.sample.1 <- readRDS("../Robjects/BC_sample1_dimred.rds")
ori.sce.sample.2 <- readRDS("../Robjects/BC_sample2_dimred.rds")

dec.sample.1 <- readRDS("../Robjects/BC_dec1_dimred.rds")
dec.sample.2 <- readRDS("../Robjects/BC_dec2_dimred.rds")

```

## Prepare the data

The first and most obvious is to subset all batches to the common “universe” of features. In this case, it is straightforward as both batches use Ensembl gene annotation; more difficult integrations will require some mapping of identifiers using the packages from Annotation Hub.

```{r}
universe <- intersect(rownames(ori.sce.sample.1), rownames(ori.sce.sample.2))
length(universe)
```

```{r}
# Subsetting the SingleCellExperiment object
sce.sample.1 <- ori.sce.sample.1[universe,]
sce.sample.2 <- ori.sce.sample.2[universe,]

# Also subsetting the variance modelling results
dec.sample.1 <- dec.sample.1[universe,]
dec.sample.2 <- dec.sample.2[universe,]

```

The second step is to rescale each batch to adjust for differences in sequencing depth between batches. The multiBatchNorm() function recomputes log-normalized expression values after adjusting the size factors for systematic differences in coverage between SingleCellExperiment objects. (Size factors only remove biases between cells within a single batch.) This improves the quality of the correction by removing one aspect of the technical differences between batches.

```{r}
rescaled <- multiBatchNorm(sce.sample.1, sce.sample.2)
sce.sample.1 <- rescaled[[1]]
sce.sample.2 <- rescaled[[2]]
```

Finally, we perform feature selection by averaging the variance components across all batches with the combineVar() function. We compute the average as it is responsive to batch-specific HVGs while still preserving the within-batch ranking of genes. This allows us to use the same strategies discussed earlier to select genes of interest. In contrast, approaches based on taking the intersection or union of HVGs across batches become increasingly conservative or liberal, respectively, with an increasing number of batches.

```{r}
combined.dec <- combineVar(dec.sample.1, dec.sample.2)
chosen.hvgs <- combined.dec$bio > 0
sum(chosen.hvgs)
```

When integrating datasets of variable composition, it is generally safer to err on the side of including more HVGs than are used in a single dataset analysis, to ensure that markers are retained for any dataset-specific subpopulations that might be present. That said, many of the signal-to-noise considerations discussed previously still apply here, so some experimentation may be necessary for best results.

## View the uncorrected data

Before we actually perform any correction, it is worth examining whether there is any batch effect in this dataset. We combine the two SingleCellExperiments and perform a PCA on the log-expression values for our selected subset of HVGs.

```{r}
# Synchronizing the metadata for cbind()ing.
rowData(sce.sample.1) <- rowData(sce.sample.2)
sce.sample.1$batch <- "1"
sce.sample.2$batch <- "2"
uncorrected <- cbind(sce.sample.1, sce.sample.2)

#run the PCA
set.seed(0010101010)
uncorrected <- runPCA(uncorrected, subset_row=chosen.hvgs)
```

We use graph-based clustering on the components to obtain a summary of the population structure. As our two sample populations should be replicates, each cluster should ideally consist of cells from both batches. However, we instead see clusters that are mostly comprised of cells from a single batch. This indicates that cells of the same type are artificially separated due to technical differences between batches.

```{r}
uncorrected.snn.gr <- buildSNNGraph(uncorrected, use.dimred="PCA")
uncorrected.clusters <- igraph::cluster_walktrap(uncorrected.snn.gr)$membership
uncorrected.tab <- table(Cluster=uncorrected.clusters, Batch=uncorrected$batch)
uncorrected.tab
```

This is supported by the t-SNE visualization where the separation between cells from different batches is consistent with the clustering results.

```{r}
set.seed(1111001)
uncorrected <- runTSNE(uncorrected, dimred="PCA")
plotTSNE(uncorrected, colour_by="batch")
```

Of course, the other explanation for batch-specific clusters is that there are cell types that are unique to each batch. The degree of intermingling of cells from different batches is not an effective diagnostic when the batches involved might actually contain unique cell subpopulations (which is not a consideration with these samples, but the same cannot be said in general). If a cluster only contains cells from a single batch, one can always debate whether that is caused by a failure of the correction method or if there is truly a batch-specific subpopulation. For example, do batch-specific metabolic or differentiation states represent distinct subpopulations? Or should they be merged together? Each batch correction algorithm will make different (and possibly inappropriate) decisions on what constitutes “shared” and “unique” populations.

## Correct the data - Mutual Nearest Neighbour

Consider a cell 'a' in batch 'A', and identify the cells in batch 'B' that are nearest neighbors to 'a' in the expression space defined by the selected features. Repeat this for a cell'b' in batch 'B', identifying its nearest neighbors in 'A'.

Mutual nearest neighbors are pairs of cells from different batches that belong in each other’s set of nearest neighbors. The reasoning is that MNN pairs represent cells from the same biological state prior to the application of a batch effect - see Haghverdi et al. (2018) for full theoretical details. Thus, the difference between cells in MNN pairs can be used as an estimate of the batch effect, the subtraction of which yields batch-corrected values.

Compared to linear regression, MNN correction does not assume that the population composition is the same or known beforehand. This is because it learns the shared population structure via identification of MNN pairs and uses this information to obtain an appropriate estimate of the batch effect. Instead, the key assumption of MNN-based approaches is that the batch effect is orthogonal to the biology in high-dimensional expression space. Violations reduce the effectiveness and accuracy of the correction, with the most common case arising from variations in the direction of the batch effect between clusters. Nonetheless, the assumption is usually reasonable as a random vector is very likely to be orthogonal in high-dimensional space.

The batchelor package provides an implementation of the MNN approach via the fastMNN() function. (Unlike the MNN method originally described by Haghverdi et al. (2018), the fastMNN() function performs PCA to reduce the dimensions beforehand and speed up the downstream neighbor detection steps.) We apply it to our two batches to remove the batch effect across the highly variable genes in chosen.hvgs. To reduce computational work and technical noise, all cells in all batches are projected into the low-dimensional space defined by the top 'd' principal components. Identification of MNNs and calculation of correction vectors are then performed in this low-dimensional space.

```{r}
set.seed(1000101001)
mnn.out <- fastMNN(sce.sample.1, sce.sample.2, d=50, k=20, subset.row=chosen.hvgs)
mnn.out
```

The function returns a SingleCellExperiment object containing corrected values for downstream analyses like clustering or visualization. Each column of mnn.out corresponds to a cell in one of the batches, while each row corresponds to an input gene in chosen.hvgs. The batch field in the column metadata contains a vector specifying the batch of origin of each cell.

The corrected matrix in the reducedDims() contains the low-dimensional corrected coordinates for all cells, which we will use in place of the PCs in our downstream analyses.

```{r}
dim(reducedDim(mnn.out, "corrected"))
```

A reconstructed matrix in the assays() contains the corrected expression values for each gene in each cell, obtained by projecting the low-dimensional coordinates in corrected back into gene expression space. We do not recommend using this for anything other than visualization.

```{r}
assay(mnn.out, "reconstructed")
```

The most relevant parameter for tuning fastMNN() is k, which specifies the number of nearest neighbors to consider when defining MNN pairs. This can be interpreted as the minimum anticipated frequency of any shared cell type or state in each batch. Increasing k will generally result in more aggressive merging as the algorithm is more generous in matching subpopulations across batches. It can occasionally be desirable to increase k if one clearly sees that the same cell types are not being adequately merged across batches.

## View the corrected data

We cluster on the low-dimensional corrected coordinates to obtain a partitioning of the cells that serves as a proxy for the population structure. If the batch effect is successfully corrected, clusters corresponding to shared cell types or states should contain cells from multiple batches. We see that all clusters contain contributions from each batch after correction, consistent with our expectation that the two batches are replicates of each other.

```{r}
mnn.snn.gr <- buildSNNGraph(mnn.out, use.dimred="corrected")
mnn.clusters <- igraph::cluster_walktrap(mnn.snn.gr)$membership
mnn.tab <- table(Cluster=mnn.clusters, Batch=mnn.out$batch)
mnn.tab
```

```{r}
set.seed(0010101010)
mnn.out <- runTSNE(mnn.out, dimred="corrected")

mnn.out$batch <- factor(mnn.out$batch)
plotTSNE(mnn.out, colour_by="batch")
```

```{r, message=FALSE}
# make dataframes
clusterTab.uncorrected <- data.frame(clusters=uncorrected.clusters, batch=uncorrected$batch, source=uncorrected$batch) %>%
  group_by(clusters,batch) %>%
  summarise(cells = n())
clusterTab.mnn <- data.frame(clusters=mnn.clusters, batch=mnn.out$batch, source=mnn.out$batch) %>%
  group_by(clusters,batch) %>%
  summarise(cells = n())

# plot bars
bar.uncorrected <- ggplot(data=clusterTab.uncorrected, 
                          aes(x=clusters,y=cells, fill=batch)) +
	theme(axis.text.x=element_text(angle = 90, hjust = 0)) +
	geom_col() +
	theme(legend.text = element_text(size = 7))
bar.mnn <- ggplot(data=clusterTab.mnn, 
              aes(x=clusters,y=cells, fill=batch)) +
	theme(axis.text.x=element_text(angle = 90, hjust = 0)) +
	geom_col() +
	theme(legend.text = element_text(size = 7))

plot_grid(bar.uncorrected, bar.mnn, ncol=1)
```

## The quickCorrect wrapper

The batchelor package has actually made this even easier for us by having a wrapper function called quickCorrect which does alot of these steps for us all at once. It performs all the steps we did to prepare the data and runs the correction. By default this is the fastMNN correction but you can change to use other correction algorithms by modifying the PARAM= argument.

1. Intersecting each batch to the universe of common features with intersectRows.

2. Applying normalization and log-transformation to the batches with multiBatchNorm.

3. Modelling the per-gene variance with modelGeneVar. If precomputed is supplied, the precomputed results are used instead.

4. Identifying highly variable genes with getTopHVGs. These genes will be used in the correction, though corrected values for all genes can be returned by setting correct.all=TRUE.

5. Applying the batch correction algorithm of choice with batchCorrect, as specified by PARAM.

```{r}
quick.corrected <- quickCorrect(ori.sce.sample.1, ori.sce.sample.2)

quick.sce <- quick.corrected$corrected
quick.sce
```

```{r}
set.seed(00101010)
quick.sce <- runTSNE(quick.sce, dimred="corrected")
quick.sce$batch <- factor(quick.sce$batch)
plotTSNE(quick.sce, colour_by="batch")
```

## Multiple Batches

The above example showed two samples or 'batches' but it will often be the case that we have many samples or batches. It is straightforward to simultaneously perform correction across >2 batches, either by having multiple levels in batch= or by providing more SingleCellExperiment objects (or even raw matrices of expression values). Lets try this out with more of the samples from our dataset.

### Load Data

This is a large list of 7 SingleCellExperiment objects from our dataset that have been processed as discussed in the previous sessions and downsampled to 500 cells per sample for processing speed.

```{r}
# ETV6_RUNX1 and PBMMC dataset
all.sce <- readRDS("../Robjects/DataIntegration_all_sce_dimred.Rds")
all.sce
```

## Exercise 1

:::exercise

Run quickCorrect on our new SingleCellExperiment object treating each individual sample as a batch.

<details><summary>Hint</summary>
 
When you use a single `sce` object you must specify what you want to use as a batch using the `batch = ` argument. See the Help page for more details.

</details>

<details><summary>Answer</summary>
```{r}
quick.corrected.all <- quickCorrect(all.sce, batch = all.sce$SampleName)

quick.sce.all <- quick.corrected.all$corrected

quick.sce.all
```
</details>

:::

## Exercise 2

:::exercise

Plot the t-SNE of your corrected data and compare it to the uncorrected data.

![uncorrected](Images/uncorrected_tsne.png)

<details><summary>Answer</summary>
```{r}
set.seed(0010101011)
quick.sce.all <- runTSNE(quick.sce.all, dimred="corrected")
quick.sce.all$batch <- factor(quick.sce.all$batch)
plotTSNE(quick.sce.all, colour_by="batch")
```
</details>

:::

When running the correction it is important to note that the metadata stored in the `colData` of the input object will not be present in the output object. We have to recreate this if we want to use it later. We can easily do this as the output object will be ordered the same as the input.

```{r}
meta <- colData(all.sce) %>%
  as.data.frame() %>%
  DataFrame()

colData(quick.sce.all) <- meta
  
```

## Correction Diagnostics

### Mixing Between Batches

Ideally, batch correction would remove the differences between batches while preserving the heterogeneity within batches. In the corrected data, cells of the same type should be intermingled and indistinguishable even if they come from different batches, while cells of different types should remain well-separated. Unfortunately, we rarely have prior knowledge of the underlying types of the cells, making it difficult to unambiguously determine whether differences between batches represent geniune biology or incomplete correction. Indeed, it could be said that all correction methods are at least somewhat incorrect, though that not preclude them from being useful.

We cluster on the low-dimensional corrected coordinates to obtain a partitioning of the cells that serves as a proxy for the population structure. If the batch effect is successfully corrected, clusters corresponding to shared cell types or states should contain cells from multiple batches. 

```{r}
all.snn.gr <- buildSNNGraph(quick.sce.all, use.dimred="corrected", k=20)
clusters.all <- igraph::cluster_walktrap(all.snn.gr)$membership
colLabels(quick.sce.all) <- factor(clusters.all)

clusterTab.all <- data.frame(clusters=clusters.all, batch=quick.sce.all$SampleName, source=quick.sce.all$SampleGroup)

all.tab <- table(Cluster=clusters.all, Batch=quick.sce.all$SampleName)
all.tab
```

```{r}
ClusterInfo.all <- data.frame(Cluster=clusters.all, Batch=quick.sce.all$SampleName, source=quick.sce.all$SampleGroup) %>%
  group_by(Cluster,Batch) %>%
  summarise(cells = n())

ggplot(data=ClusterInfo.all, aes(x=Cluster,y=cells, fill=Batch)) +
	theme(axis.text.x=element_text(angle = 90, hjust = 0)) +
	geom_col() +
	theme(legend.text = element_text(size = 7))

```

The OSCA book favors a qualitative approach to assessing the degree of mixing between clusters where they compute the variance in the log-normalized abundances across batches for each cluster. A highly variable cluster has large relative differences in cell abundance across batches; this may be an indicator for incomplete batch correction, e.g., if the same cell type in two batches was not combined into a single cluster in the corrected data. We can then focus our attention on these clusters to determine whether they might pose a problem for downstream interpretation. Of course, a large variance can also be caused by genuinely batch-specific populations, so some prior knowledge about the biological context is necessary to distinguish between these two possibilities.

```{r}
var <- clusterAbundanceVar(all.tab)

# Also printing the percentage of cells in each cluster in each batch:
percent <- t(t(all.tab)/colSums(all.tab)) * 100 
df <- DataFrame(Batch=unclass(percent), var=var)
df[order(df$var, decreasing=TRUE),]

```

### Preserving Biological Heterogeneity

Another useful diagnostic check is to compare the pre-correction clustering of each batch to the clustering of the same cells in the corrected data. Accurate data integration should preserve population structure within each batch as there is no batch effect to remove between cells in the same batch. This check complements the previously mentioned diagnostics that only focus on the removal of differences between batches. Specifically, it protects us against scenarios where the correction method simply aggregates all cells together, which would achieve perfect mixing but also discard the biological heterogeneity of interest. Lets go back to our simple two sample example to look at some of the ways we can investigate.

```{r}
table(colLabels(ori.sce.sample.1))
```

```{r}
table(colLabels(ori.sce.sample.2))
```

Ideally, we should see a many-to-1 mapping where the post-correction clustering is nested inside the pre-correction clustering. This indicates that any within-batch structure was preserved after correction while acknowledging that greater resolution is possible with more cells. We quantify this mapping using the nestedClusters() function from the bluster package, which identifies the nesting of post-correction clusters within the pre-correction clusters. Well-nested clusters have high `max` values, indicating that most of their cells are derived from a single pre-correction cluster.

```{r}
tab.sample.1 <- nestedClusters(ref=paste("before", colLabels(ori.sce.sample.1)),
    alt=paste("after", mnn.clusters[mnn.out$batch==1]))
tab.sample.1$alt.mapping
```

We can visualize this mapping for the samples. Ideally, each row should have a single dominant entry close to unity. Horizontal stripes are more concerning as these indicate that multiple pre-correction clusters were merged together, though the exact level of concern will depend on whether specific clusters of interest are gained or lost. In practice, more discrepancies can be expected even when the correction is perfect, due to the existence of closely related clusters that were arbitrarily separated in the within-batch clustering.

```{r}
# For the first batch:
heat.sample.1 <- pheatmap(tab.sample.1$proportions, cluster_row=FALSE, cluster_col=FALSE,
                   main="Sample 1 comparison", silent=TRUE)

# For the second batch:
tab.sample.2 <- nestedClusters(ref=paste("before", colLabels(ori.sce.sample.2)),
                        alt=paste("after", mnn.clusters[mnn.out$batch==2]))
heat.sample.2 <- pheatmap(tab.sample.2$proportions, cluster_row=FALSE, cluster_col=FALSE,
                   main="Sample 2 comparison", silent=TRUE)

gridExtra::grid.arrange(heat.sample.1[[4]], heat.sample.2[[4]])
```

We use the adjusted Rand index to quantify the agreement between the clusterings before and after batch correction. Larger indices are more desirable as this indicates that within-batch heterogeneity is preserved, though this must be balanced against the ability of each method to actually perform batch correction.

```{r}
ri.sample.1 <- pairwiseRand(mnn.clusters[mnn.out$batch==1], colLabels(ori.sce.sample.1), mode="index")
ri.sample.1
```

```{r}
ri.sample.2 <- pairwiseRand(mnn.clusters[mnn.out$batch==2], colLabels(ori.sce.sample.2), mode="index")
ri.sample.2
```

We can also break down the ARI into per-cluster ratios for more detailed diagnostics (Figure 2.2). For example, we could see low ratios off the diagonal if distinct clusters in the within-batch clustering were incorrectly aggregated in the merged clustering. Conversely, we might see low ratios on the diagonal if the correction inflated or introduced spurious heterogeneity inside a within-batch cluster.

```{r}
# For the first batch.
tab <- pairwiseRand(colLabels(ori.sce.sample.1), mnn.clusters[mnn.out$batch==1])
heat.1 <- pheatmap(tab, cluster_row=FALSE, cluster_col=FALSE,
    col=rev(viridis::magma(100)), main="Sample 1 probabilities", silent=TRUE)

# For the second batch.
tab <- pairwiseRand(colLabels(ori.sce.sample.2), mnn.clusters[mnn.out$batch==2])
heat.2 <- pheatmap(tab, cluster_row=FALSE, cluster_col=FALSE,
    col=rev(viridis::magma(100)), main="Sample 2 probabilities", silent=TRUE)

gridExtra::grid.arrange(heat.1[[4]], heat.2[[4]])

```

### MNN specific test

For fastMNN(), one useful diagnostic is the proportion of variance within each batch that is lost during MNN correction. Specifically, this refers to the within-batch variance that is removed during orthogonalization with respect to the average correction vector at each merge step. This is returned via the lost.var field in the metadata of mnn.out, which contains a matrix of the variance lost in each batch (column) at each merge step (row).

```{r}
metadata(mnn.out)$merge.info$lost.var
```

Large proportions of lost variance (>10%) suggest that correction is removing genuine biological heterogeneity. This would occur due to violations of the assumption of orthogonality between the batch effect and the biological subspace (Haghverdi et al. 2018). In this case, the proportion of lost variance is smaller, indicating that non-orthogonality is not to much of a major concern.

## Session information

```{r sessionInfo_dsi_PBMMC_ETV6-RUNX1}
sessionInfo()
```