---
title: "Introduction to single-cell RNA-seq analysis"
subtitle: "Cluster marker genes"
date: "May 2021"
author: "Ashley Sawle, Stephane Ballereau, Zeynep Kalender Atak, Hugo Tavares"
output:
  html_notebook:
    toc: yes
    toc_depth: 2
    css: ../css/boxes.css
---


```{r setup, echo=FALSE, include=FALSE, message=FALSE, purl=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, cache=TRUE)
# knitr::opts_chunk$set(fig.width=7, fig.height=7)
# knitr::opts_chunk$set(dev="CairoPNG")
set.seed(123)
```


# Introduction

In order to aid the interpretation of the clustering results, it is helpful to
identify genes that are contributing to the separation of cells into those
clusters.

The main approach to achieve this, is to identify genes that are differently
expressed between clusters. These may be, for example, exclusively expressed in
a single cluster or perhaps differentiate between a few different clusters.
There are different methods to identify differentially expressed genes, focusing
on differences between the mean expression level, the differences in the ranking
of genes by expression between clusters or the differences between clusters of
the proportions of cells expressing a gene.

Our main objective in this section is to cover some of the methods that can be
used to achieve this goal, and obtain a summary table of results.

As always, the [OSCA chapter on marker detection](https://bioconductor.org/books/release/OSCA.basic/marker-detection.html) 
contains additional detail and advice.

# Setup

Before we start, let's load our packages and read our data in. 

```{r packages, results='hide', message=FALSE, warning=FALSE}
library(scater)
library(scran)
library(pheatmap)
library(tidyverse)
```

We will load two _SingleCellExperiment_ objects generated in previous sections:

* The SCE with logcounts normalised within each batch generated in the [Dataset
Integration and Batch Correction](07_DatasetIntegration.html) section
* The SCE with batch-corrected data integration using the `batchelor::fastMNN()`
algorithm, including clusters covered in the [Clustering](08_Clustering_Materials.nb.html)
section.  

```{r}
uncorrected <- readRDS("../Robjects/DataIntegration_all_sce_dimred.Rds")
rownames(uncorrected) <- uniquifyFeatureNames(rownames(uncorrected), rowData(uncorrected)$Symbol)

# data corrected using batch integration with mutual nearest neighbours
corrected <- readRDS("../Robjects/Caron_clustering_material.rds")
rownames(corrected) <- uniquifyFeatureNames(rownames(corrected), rowData(corrected)$Symbol)
```

Note that we also replace the gene names (rownames) of our objects to use common
gene symbols instead of Ensembl IDs. Using the `uniquifyFeatureNames()` function
makes this safely, avoiding duplicated gene symbols.

As a reminder, the `corrected` object contains the cell cluster assignments
obtained in a `colData` column called "label":

```{r}
# visualise cluster assignments on the corrected data
plotTSNE(corrected, 
         colour_by = "label", 
         text_by = "label")
```

It will also be helpful for data visualisation purposes to copy the cluster
assignment in the corrected data to our uncorrected (normalised) SCE object:

```{r}
# first make sure that the cell names are in the same order
all(colnames(uncorrected) == colnames(corrected))
colLabels(uncorrected) <- colLabels(corrected)
```


# Identifying cluster marker genes

Although we have defined our clusters based on the batch-corrected expression
values, these should **not** be used for for gene-based analyses like marker
gene detection.

Instead, we should use the uncorrected (normalised) expression values for
differential expression between clusters.

The reason for this is that batch correction algorithms do not necessarily
preserve the magnitude (or even direction) of the differences in gene expression
when attempting to integrate multiple batches. For example, the cosine
normalization used by the "Mutual Nearest Neighbours" algorithm (`fastMNN()`)
shrinks the magnitude of the expression values so that the computed log-fold
changes have no obvious interpretation. Of greater concern is the possibility
that the correction introduces artificial agreement across batches.

In our case, the valid assay to use in differential analysis tests is the
logcounts obtained from the deconvolution method (using `computePooledFactors` &
`logNormCounts`).

## Pairwise comparisons for scoring potential marker genes

The basic approach for marker gene identification across clusters is to perform
statisticals tests for each gene between every pair of clusters based on gene
expression. We use `scoreMarkers()` function to do this. In addition to
computing various scores, the function can take account of known batch effects.
The function outputs a list of DataFrames, with one DataFrame of scores for each
cluster. However, note that the blocking assumes that each pair of clusters is 
present in at least one of the blocks. If there are two clusters which are not
both present in at least one block (in our cases Samples), then that pairwise
comparison will by necessity be omitted.

```{r}
markers <- scoreMarkers(uncorrected, 
                        groups = factor(uncorrected$label), 
                        block =uncorrected$SampleName)

head(markers[["8"]], n=2)
```

This DataFrame contains the results for cluster 8. The first four columns contain summary 
statistics:

* **self.average** - the mean log-expression in the cluster of interest  
* **other.average** - the grand mean across all other clusters  
* **self.detected** - the proportion of cells with detected expression in the cluster of interest  
* **other.detected** - the mean detected proportion across all other clusters.  

The remaining columns contain summaries of three scores from the pairwise comparisons. The three scores are:

* **logFC.cohen** - "Cohens *d*" - this is the log fold change of mean gene
expression that has been standardized by scaling by the average of the standard
deviation across the groups. This can be interpreted in a similar way to log
fold change in that a positive value indicates upregulation in the cluster of
interest.    
* **AUC** - "Area Under the Curve" - this quantifies the ability to distinguish
between two gene expression distributions. It can be interpreted as the
likelihood that any random cell in the cluster of interest will have a higher
expression of the gene than any random cell in the other cluster. It ranges from
0 to 1, where 1 can be interpreted as upregulation, 0 downregulation, and 0.5 as
no difference.
* **logFC.detected** - this is the log fold change in the proportion of cells in
which the gene is detected in the cluster of interest versus the proportion of
cells in which the gene is detected in the other cluster. Positive values
indicate that the gene is detected in a more cells in the cluster of interest
than the other cluster. Note this takes no account of the magnitude of the gene
expression.

More detail on the differences between these effect size scores can be found in
the ["Advanced" Marker detection chapter of the OSCA
book](http://bioconductor.org/books/3.15/OSCA.advanced/marker-detection-redux.html#properties-of-each-effect-size).

Whilst all the pairwise scores can be retrieved by adding the argument
`full.stats=TRUE` to the `scoreMarkers` command, by default `scoreMarkers`
returns 5 summary statistics for each scrore:

* **mean.X** - this is the mean of the score across all pairwise comparisons. It
gives the relative expression of the gene versus the average of the other
clusters.  
* **min.X** - this is the minimum score obtained across all pairwise
comparisons. This is the most stringent summary statistic for detecting
upregulated genes, if the score is high, then the gene is upregulated in the
cluster of interest relative to *all* other clusters. Conversely, if the minimum
is low, then the gene is downregulated relative to *at least one* other cluster.  
* **median.X** - this is the median of the score across all pairwise
comparisons. It is more robust to outliers than the mean. Whilst this is
beneficial in avoiding the situation where the effect size is very large in only
a small number of comparisons, it may conversely be detrimental to marker gene
selection if there are small number of clusters with very similar effect sizes,
as these will be effectively ignored.
* **max.X** - this is the maximum score obtained across all pairwise
comparisons. This is the least stringent summary statistic for detecting
upregulated genes as a high score ony indicates that the gene is upregulated in
the cluster of interest relative to *at least one* other clusters. Conversely,
if the maximum is low, then the gene is downregulated relative to *all* other
clusters.
* **rank.X** - This is the minimum ranking ("min-rank") of that gene by that
score across all clusters. For each pairwise comparison the genes are ranked
according to the score, this summary provides the lowest rank for that gene
across all comparisons. Essentially, a gene with a low "min-rank" will be one of
the best genes according to that score in at least one comparison.

The choice of the summary used for ranking will effect the stringency of the
selection. See the [the OSCA books "Basic" chapter on Marker gene
detection](http://bioconductor.org/books/3.15/OSCA.basic/marker-detection.html#summarizing-pairwise-effects)
for further discussion of these different summaries. In general the mean and
median make reasonable defaults for most applications. In practice, the minimum
and maximum are most helpful for diagnosing discrepancies between the mean and
median, rather than being used directly for ranking.

Selecting genes based on a given min-rank, say *5*, is useful as it will
generate a list of genes that is the union of genes with a rank of 5 or less for
any pairwise comparison.This will ensure we get at least 5 genes that
distinguish the cluster of interest from all other clusters.

For exmple using the min-rank for Cohen's *d* on cluster 11 yeilds 19 marker
genes:

```{r}
topRanked <- markers[["11"]] %>%
  as.data.frame() %>% 
  select(contains("cohen")) %>% 
  filter(rank.logFC.cohen <= 5) %>% 
  arrange(rank.logFC.cohen)
topRanked
```

We can then plot these on a heatmap to visualize the gene expression differences:

```{r fig.height=6, fig.width=6}
plotGroupedHeatmap(uncorrected, 
                   features=rownames(topRanked), 
                   group="label", 
                   block="SampleName",
                   center=TRUE, 
                   zlim=c(-3, 3))
```

We can plot the distribution of  expression of sepicfic genes as we have
previously, e.g. FCGR3A (CD16), a monocyte marker:

```{r}
plotTSNE(corrected, 
         by_exprs_values = "reconstructed", 
         colour_by = "FCGR3A",
         text_by = "label")
```

# Adjusting the log-fold change

The AUC and Cohen's *d* scores incorporate both the gene expression differences
between the clusters and the vairiance in gene expression scores within each
cluster. If a gene with low variance it is possible that it will be ranked
highly even if the magnitude of the difference between the clusters is low.
These genes will not necessarily make good marker genes. It may therefore be
desirable to favour the detection of genes with larger log-fold changes. A
log-fold change threshold can be set using the `lfc=` argument in score markers.

Consider for example cluster 6. The gene RPS18, a ribosomal protein gene, is
identified as a good marker for this clusters based on a min-rank for Cohen's
*d* of 2.

```{r}
topRanked <- markers[["6"]] %>%
  as.data.frame() %>% 
  select(contains("cohen")) %>% 
  filter(rank.logFC.cohen <= 5) %>% 
  arrange(rank.logFC.cohen)
topRanked
```

However, if we are interested in finding strong positive markers, a look at the
expression shows that it would not really be a good positive marker as
fold-change difference between it and other clusters is not great.

```{r}
plotExpression(uncorrected, 
               features = "RPS18", 
               x = "label")
```
We can select for strong positive markers by setting an `lfc` threshold of 2:

```{r}
markerslfc <- scoreMarkers(uncorrected, 
                        groups = factor(uncorrected$label), 
                        block =uncorrected$SampleName,
                        lfc=2)
```

RPS18 no longer appears in the candidate cluster marker genes list by min-rank of Cohen's *d*.

```{r}
topRanked <- markerslfc[["6"]] %>%
  as.data.frame() %>% 
  select(contains("cohen")) %>% 
  filter(rank.logFC.cohen <= 5) %>% 
  arrange(rank.logFC.cohen)
topRanked
```

In fact it's min-rank for Cohen's *d* has dropped to:

```{r}
markerslfc[["6"]]["RPS18", "rank.logFC.cohen"]
```


:::note

From the OSCA book:  

> Given that scoreMarkers() already reports effect sizes, it is tempting to take
the next step and obtain p-values for the pairwise comparisons. Unfortunately,
the p-values from the relevant tests cannot be reliably used to reject the null
hypothesis. This is because DE analysis is performed on the same data used to
obtain the clusters, which represents “data dredging” (also known as fishing or
data snooping). The hypothesis of interest - are there differences between
clusters? - is formulated from the data, so we are more likely to get a positive
result when we re-use the data set to test that hypothesis.

More details [here](http://bioconductor.org/books/3.15/OSCA.advanced/marker-detection-redux.html#from-data-snooping)

:::

# Session information

<details><summary>`sessionInfo()`</summary>

```{r, purl=FALSE}
sessionInfo()
```

</details>